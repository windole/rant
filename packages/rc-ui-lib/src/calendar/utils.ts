/* eslint-disable no-nested-ternary */
export function compareMonth(date1: Date, date2: Date): number {
  const year1 = date1.getFullYear();
  const year2 = date2.getFullYear();

  if (year1 === year2) {
    const month1 = date1.getMonth();
    const month2 = date2.getMonth();
    return month1 === month2 ? 0 : month1 > month2 ? 1 : -1;
  }

  return year1 > year2 ? 1 : -1;
}

export function compareDay(day1: Date, day2: Date): number {
  const compareMonthResult = compareMonth(day1, day2);

  if (compareMonthResult === 0) {
    const date1 = day1.getDate();
    const date2 = day2.getDate();
    return date1 === date2 ? 0 : date1 > date2 ? 1 : -1;
  }

  return compareMonthResult;
}

export const cloneDate = (date: Date): Date => new Date(date);

export const cloneDates = (dates: Date | Date[]): Date | Date[] =>
  Array.isArray(dates) ? dates.map(cloneDate) : cloneDate(dates);

export function getDayByOffset(date: Date, offset: number): Date {
  const cloned = cloneDate(date);
  cloned.setDate(cloned.getDate() + offset);
  return cloned;
}

export const getPrevDay = (date: Date): Date => getDayByOffset(date, -1);
export const getNextDay = (date: Date): Date => getDayByOffset(date, 1);
export const getToday = (): Date => {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  return today;
};

export function calcDateNum(date: [Date, Date]): number {
  const day1 = date[0].getTime();
  const day2 = date[1].getTime();
  return (day2 - day1) / (1000 * 60 * 60 * 24) + 1;
}

export const getMonthEndDay = (year: number, month: number): number =>
  32 - new Date(year, month - 1, 32).getDate();

const calendarLang = {
  end: '结束',
  start: '开始',
  title: '日期选择',
  confirm: '确定',
  startEnd: '开始/结束',
  weekdays: ['日', '一', '二', '三', '四', '五', '六'],
  monthTitle: (year: number, month: number) => `${year}年${month}月`,
  rangePrompt: (maxRange: number) => `最多选择 ${maxRange} 天`,
};

export const t = (key: string, ...args): string => {
  if (args.length) {
    return calendarLang[key](...args);
  }
  return calendarLang[key];
};

export const formatMonthTitle = (date: Date): string =>
  t('monthTitle', date.getFullYear(), date.getMonth() + 1);
