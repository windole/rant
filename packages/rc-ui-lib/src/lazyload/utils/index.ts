import { ListenEvent } from '../PropsType';

/* eslint-disable prefer-rest-params */
const inBrowser = typeof window !== 'undefined';

export const DEFAULT_URL =
  'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';

// 需要监听的事件
export const DEFAULT_EVENTS: ListenEvent[] = [
  'scroll',
  'wheel',
  'resize',
  'mousewheel',
  'animationend',
  'transitionend',
  'touchmove',
];

export const hasIntersectionObserver =
  inBrowser &&
  'IntersectionObserver' in window &&
  'IntersectionObserverEntry' in window &&
  'intersectionRatio' in window.IntersectionObserverEntry.prototype;

export const modeType = {
  event: 'event',
  observer: 'observer',
};

export const call = (fn): void => fn && fn();

export function setImage(
  element: HTMLMediaElement | HTMLIFrameElement | HTMLImageElement | HTMLElement,
  imagePath: string,
): void {
  if (
    element instanceof HTMLIFrameElement ||
    element instanceof HTMLImageElement ||
    element instanceof HTMLMediaElement
  ) {
    element.src = imagePath;
  } else {
    element.style.backgroundImage = `url('${imagePath}')`;
  }
}

export function loadImage(imagePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.src = imagePath;
    img.onload = () => resolve(imagePath);
    img.onerror = (err) => reject(err);
  });
}

export function addCssClassName(
  element: HTMLImageElement | HTMLElement,
  cssClassName: string,
): void {
  if (!element.className.includes(cssClassName)) {
    element.className += ` ${cssClassName}`;
  }
}

export function throttle(fn, threshold = 250, scope?) {
  let last;
  let deferTimer;
  return () => {
    // @ts-ignore
    const context = scope || this;

    const now = +new Date();
    const args = arguments;
    if (last && now < last + threshold) {
      // hold on to it
      clearTimeout(deferTimer);
      deferTimer = setTimeout(() => {
        last = now;
        fn.apply(context, args);
      }, threshold);
    } else {
      last = now;
      fn.apply(context, args);
    }
  };
}

export function debounce(func, wait, immediate?) {
  let timeout;
  let args;
  let context;
  let timestamp;
  let result;

  const later = function later() {
    const last = +new Date() - timestamp;

    if (last < wait && last >= 0) {
      timeout = setTimeout(later, wait - last);
    } else {
      timeout = null;
      if (!immediate) {
        result = func.apply(context, args);
        if (!timeout) {
          context = null;
          args = null;
        }
      }
    }
  };

  return function debounced() {
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    context = this;
    args = arguments;
    timestamp = +new Date();

    const callNow = immediate && !timeout;
    if (!timeout) {
      timeout = setTimeout(later, wait);
    }

    if (callNow) {
      result = func.apply(context, args);
      context = null;
      args = null;
    }

    return result;
  };
}
