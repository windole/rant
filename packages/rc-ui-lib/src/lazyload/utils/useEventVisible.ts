import { useEffect, useRef, useState } from 'react';
import { debounce as debounceFunc, DEFAULT_EVENTS, throttle as throttleFunc } from '.';
import { EventOptions } from '../PropsType';

// try to handle passive events
let passiveEventSupported = false;
try {
  const opts = Object.defineProperty({}, 'passive', {
    get() {
      passiveEventSupported = true;
    },
  });
  window.addEventListener('test', null, opts);
  // eslint-disable-next-line no-empty
} catch (e) {}

const passiveEvent = passiveEventSupported ? { capture: false, passive: true } : false;

const useEventVisible = (
  eventOptions: EventOptions,
): [React.MutableRefObject<HTMLDivElement | HTMLImageElement>, boolean] => {
  const {
    listenEvents = DEFAULT_EVENTS,
    offset,
    scrollContainer,
    debounce,
    throttle,
  } = eventOptions;
  // 是否可见
  const [visible, setVisible] = useState(false);
  // // 包裹容器
  const containerRef = useRef<HTMLDivElement | HTMLImageElement>(null);

  useEffect(() => {
    // 检查函数
    let checkVisible = () => {
      // 如果容器不存在则不计算
      if (!containerRef.current) return;
      // 获取当前组件位置
      const { top, left, width, height } = containerRef.current?.getBoundingClientRect();
      // 获取滚动容器位置
      const {
        top: parentTop,
        left: parentLeft,
        width: parentWidth,
        height: parentHeight,
      } = scrollContainer.getBoundingClientRect();
      // 计算屏幕高度和宽度
      const windowInnerHeight = window.innerHeight || document.documentElement.clientHeight;
      const windowInnerWidth = window.innerWidth || document.documentElement.clientWidth;
      // 计算滚动容器和屏幕可视区的相交区域
      const intersectionTop = Math.min(parentTop, 0);
      const intersectionLeft = Math.min(parentLeft, 0);

      const intersectionHeight =
        Math.min(windowInnerHeight, parentTop + parentHeight) - intersectionTop;
      const intersectionWidth =
        Math.min(windowInnerWidth, parentLeft + parentWidth) - intersectionLeft;
      // 计算组件距离可视区的高度
      const offsetTop = top - intersectionTop;
      const offsetLeft = left - intersectionLeft;
      // 偏移量计算[top, left]
      const offsets = Array.isArray(offset) ? offset : [offset, offset];
      // 通过上述距离判断组件是否在可见区域
      const isVisible =
        offsetTop - offsets[0] <= intersectionHeight &&
        offsetTop + height + offsets[0] >= 0 &&
        offsetLeft - offsets[1] <= intersectionWidth &&
        offsetLeft + width + offsets[1] >= 0;

      setVisible(isVisible);
      // 如果可见取消当前组件的所有的监听
      if (isVisible) {
        listenEvents.forEach((event) => {
          scrollContainer?.removeEventListener(event, checkVisible);
        });
      }
    };

    // 使用防抖节流增强checkVisible函数，这里优先防抖，其次节流
    if (debounce) {
      // 防抖函数
      checkVisible = debounceFunc(checkVisible, debounce);
    } else if (throttle) {
      // 节流函数
      checkVisible = throttleFunc(checkVisible, throttle);
    }
    // 批量监听事件
    listenEvents.forEach((event) => {
      scrollContainer?.addEventListener(event, checkVisible, passiveEvent);
    });

    // 自动执行一次
    checkVisible();

    return () => {
      listenEvents.forEach((event) => {
        scrollContainer?.removeEventListener(event, checkVisible, passiveEvent);
      });
    };
  }, [scrollContainer]);
  return [containerRef, visible];
};

export default useEventVisible;
