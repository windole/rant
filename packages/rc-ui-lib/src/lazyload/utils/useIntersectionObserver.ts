import { useCallback, useEffect, useState } from 'react';

const useIntersectionObserver = (
  observerOptions: IntersectionObserverInit,
  handleIntersection: IntersectionObserverCallback,
): ((node: Element) => void) => {
  const [element, setElement] = useState<Element>(null);

  const handleElement = useCallback((node: Element) => {
    setElement(node);
  }, []);

  useEffect(() => {
    if (!element) return;
    const observer = new IntersectionObserver(handleIntersection, observerOptions);
    observer.observe(element);
    // eslint-disable-next-line consistent-return
    return () => observer && observer.disconnect && observer.disconnect();
  }, [handleIntersection, observerOptions, element]);
  return handleElement;
};

export default useIntersectionObserver;
