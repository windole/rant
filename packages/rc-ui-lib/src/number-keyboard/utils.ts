import { KeyConfig, NumberKeyboardProps } from './PropsType';

const genBasicKeys = (props: NumberKeyboardProps) => {
  const keys: KeyConfig[] = Array(9)
    .fill('')
    .map((_, i) => ({ text: i + 1 }));

  if (props.randomKeyOrder) {
    keys.sort(() => (Math.random() > 0.5 ? 1 : -1));
  }

  return keys;
};

export const genDefaultKeys = (props: NumberKeyboardProps): KeyConfig[] => [
  ...genBasicKeys(props),
  { text: props.extraKey as string, type: 'extra' },
  { text: 0 },
  {
    text: props.showDeleteKey ? props.deleteButtonText : '',
    type: props.showDeleteKey ? 'delete' : '',
  },
];

export const genCustomKeys = (props: NumberKeyboardProps) => {
  const keys = genBasicKeys(props);
  const { extraKey } = props;
  const extraKeys = Array.isArray(extraKey) ? extraKey : [extraKey];

  if (extraKeys.length === 1) {
    keys.push({ text: 0, wider: true }, { text: extraKeys[0], type: 'extra' });
  } else if (extraKeys.length === 2) {
    keys.push(
      { text: extraKeys[0], type: 'extra' },
      { text: 0 },
      { text: extraKeys[1], type: 'extra' },
    );
  }

  return keys;
};
