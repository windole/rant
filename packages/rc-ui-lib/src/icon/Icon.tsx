import React, { useContext, useMemo } from 'react';
import classnames from 'classnames';
import { addUnit } from '../utils';
import { IconProps } from './PropsType';
import Badge from '../badge';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

function isImage(name?: string): boolean {
  return name ? name.indexOf('/') !== -1 : false;
}

const Icon: React.FC<IconProps> = (props) => {
  const { iconPrefix } = useContext(ConfigProviderContext);

  const { dot, tag, name, className, onClick, onTouchStart } = props;

  const imageIcon = isImage(name);

  const classPrefix = useMemo(
    () => props.classPrefix || iconPrefix,
    [props.classPrefix, iconPrefix],
  );

  return (
    <Badge
      dot={dot}
      tag={tag}
      className={classnames(className, classPrefix, imageIcon ? '' : `${classPrefix}-${name}`)}
      style={{
        color: props.color,
        fontSize: addUnit(props.size),
        ...props.style,
      }}
      onClick={onClick}
      onTouchStart={onTouchStart}
      {...props.badge}
    >
      {props?.children}
      {imageIcon && <img className={classnames('rc-icon__image')} src={name} alt={name} />}
    </Badge>
  );
};

Icon.defaultProps = {
  tag: 'i',
};

export default Icon;
