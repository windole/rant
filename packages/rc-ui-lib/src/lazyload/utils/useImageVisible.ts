import { useCallback, useState } from 'react';
import { addCssClassName, loadImage, setImage } from '.';
import { IntersectionObserverEntryType, ObserverOptions } from '../PropsType';
import useIntersectionObserver from './useIntersectionObserver';

const defaultObserverOptions = {
  threshold: [0, 1],
};

const useImageVisible = (
  observerOptions: ObserverOptions = defaultObserverOptions,
  errorImage: string,
): [(node: Element) => void, boolean] => {
  const [isLoaded, setIsLoaded] = useState(false);

  const handleIntersection = useCallback(
    (entries: IntersectionObserverEntryType[]) => {
      entries
        .filter((entry) => entry.isIntersecting)
        .forEach((entry) => {
          const target = entry.target as HTMLImageElement | HTMLElement;
          const url = target.dataset.src || '';
          loadImage(url)
            .catch(() => {
              setImage(target, errorImage);
              addCssClassName(target, 'rc-lazyload__error');
              target.setAttribute('lazy', 'error');
            })
            .then((imagePath: string) => {
              setImage(target, imagePath);
              setIsLoaded(true);
              addCssClassName(target, 'rc-lazyload__loaded');
              target.setAttribute('lazy', 'loaded');
              target.removeAttribute('data-src');
            });
        });
    },
    [setIsLoaded, isLoaded],
  );
  const ref = useIntersectionObserver(observerOptions, handleIntersection);
  return [ref, isLoaded];
};

export default useImageVisible;
