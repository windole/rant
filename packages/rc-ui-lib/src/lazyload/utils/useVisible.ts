import { useCallback, useEffect, useState } from 'react';
import { addCssClassName } from '.';
import useIntersectionObserver from './useIntersectionObserver';

const defaultObserverOptions = {
  threshold: [0, 1],
};

const useVisible = (
  initialVisible: boolean,
  observerOptions = defaultObserverOptions,
): [(node: Element) => void, boolean] => {
  const [visible, setVisible] = useState(initialVisible);

  const forceVisible = () => {
    setVisible(true);
  };

  useEffect(() => {
    if (initialVisible) {
      forceVisible();
    }
  }, [initialVisible]);

  const handleIntersection = useCallback(
    (entries) => {
      entries
        .filter((entry) => entry.isIntersecting)
        .forEach((entry) => {
          const target = entry.target as HTMLElement;
          addCssClassName(target, 'rc-lazyload__loaded');
          setVisible(true);
        });
    },
    [setVisible, visible],
  );
  const ref = useIntersectionObserver(observerOptions, handleIntersection);
  return [ref, visible];
};

export default useVisible;
