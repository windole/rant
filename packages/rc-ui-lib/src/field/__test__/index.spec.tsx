import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { sleep } from '../../../tests/utils';
import Icon from '../../icon';
import { Field } from '..';

describe('Field', () => {
  let wrapper;
  afterEach(() => {
    wrapper.unmount();
    jest.clearAllMocks();
  });

  it('should update value when after inputting', () => {
    const onChange = jest.fn();
    wrapper = mount(<Field onChange={onChange} />);
    const mockedEvent = {
      preventDefault: jest.fn(),
      target: { value: 'world' },
    };
    wrapper.find('input').simulate('change', mockedEvent);
    wrapper.update();
    expect(onChange).toHaveBeenCalledWith('world');
  });

  it('should emit onClickInput event when input is clicked', () => {
    const onClickInput = jest.fn();
    wrapper = mount(<Field onClickInput={onClickInput} />);
    wrapper.find('input').simulate('click');
    expect(onClickInput).toHaveBeenCalled();
  });

  it('should emit click-input event when using input slot', () => {
    const onClickInput = jest.fn();
    wrapper = mount(
      <Field onClickInput={onClickInput}>
        <input />
      </Field>,
    );
    wrapper.find('.rc-field__control').simulate('click');
    expect(onClickInput).toHaveBeenCalled();
  });

  it('should onClickLeftIcon event when left icon is clicked', () => {
    const onClickLeftIcon = jest.fn();
    wrapper = mount(
      <Field leftIcon="contact" onClickLeftIcon={onClickLeftIcon} iconPrefix="my-icon" />,
    );
    wrapper.find('.rc-field__left-icon').simulate('click');
    expect(onClickLeftIcon).toHaveBeenCalled();
  });

  it('should emit onClickRightIcon event when right icon is clicked', () => {
    const onClickRightIcon = jest.fn();
    wrapper = mount(<Field rightIcon="search" onClickRightIcon={onClickRightIcon} />);
    wrapper.find('.rc-field__right-icon').simulate('click');
    expect(onClickRightIcon).toHaveBeenCalled();
  });

  it('should format input value when type is number', () => {
    const onChange = jest.fn();
    wrapper = mount(<Field type="number" onChange={onChange} />);
    const input = wrapper.find('input');
    input.simulate('change', { preventDefault: jest.fn(), target: { value: '1' } });
    expect(onChange).toHaveBeenCalledWith('1');

    input.simulate('change', { preventDefault: jest.fn(), target: { value: '2' } });
    expect(onChange).toHaveBeenCalledWith('2');

    input.simulate('change', { preventDefault: jest.fn(), target: { value: '123abc' } });
    expect(onChange).toHaveBeenCalledWith('123');
  });

  it('should format input value when type is digit', () => {
    const onChange = jest.fn();
    wrapper = mount(<Field type="digit" value="" onChange={onChange} />);

    const input = wrapper.find('input');

    input.simulate('change', { preventDefault: jest.fn(), target: { value: '1' } });
    expect(onChange).toHaveBeenCalledWith('1');

    input.simulate('change', { preventDefault: jest.fn(), target: { value: '2' } });
    expect(onChange).toHaveBeenCalledWith('2');

    input.simulate('change', { preventDefault: jest.fn(), target: { value: '123abc' } });
    expect(onChange).toHaveBeenCalledWith('123');
  });

  it('should render textarea when type is textarea', async () => {
    wrapper = mount(<Field type="textarea" autosize />);
    await sleep(0);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should autosize textarea field', async () => {
    wrapper = mount(<Field type="textarea" autosize />);

    const value = '1'.repeat(20);
    const textarea = wrapper.find('.rc-field__control');

    await wrapper.setProps({ value });
    expect(textarea.text()).toEqual(value);
  });

  it('should allow autosize prop be be an object', async () => {
    window.scrollTo = jest.fn();
    const autosize = {
      maxHeight: 100,
      minHeight: 50,
    };

    wrapper = mount(<Field type="textarea" autosize={autosize} />);
    const textarea = wrapper.find('.rc-field__control');
    await sleep(0);
    expect(textarea.getDOMNode().style.height).toEqual('50px');
  });

  it('should onFocus call correctly', () => {
    const onFocus = jest.fn();
    wrapper = mount(<Field type="text" onFocus={onFocus} />);
    wrapper.find('input').simulate('focus');
    expect(onFocus).toHaveBeenCalled();
  });

  it('should onBlur called correctly', () => {
    const onBlur = jest.fn();
    wrapper = mount(<Field type="text" onBlur={onBlur} />);
    wrapper.find('input').simulate('blur');
    expect(onBlur).toHaveBeenCalled();
  });

  it('should limit maxlength of input value when using maxlength prop', async () => {
    const onChange = jest.fn();
    wrapper = mount(<Field type="number" maxlength={3} value="1234" onChange={onChange} />);
    const input = wrapper.find('input');
    await input.simulate('change');
    expect(onChange).toHaveBeenCalledWith('123');
  });

  it('should render clear icon when using clearable prop', async () => {
    const onFocus = jest.fn();
    wrapper = mount(<Field clearable type="text" value="test" onFocus={onFocus} />);
    expect(wrapper.find('.rc-field__clear').exists()).toBeFalsy();
    const input = wrapper.find('input');
    await input.simulate('focus');
    expect(wrapper.find('.rc-field__clear').exists()).toBeTruthy();
  });

  it('should always render clear icon when clear-trigger prop is always', async () => {
    wrapper = mount(<Field clearable type="text" value="test" />);
    expect(wrapper.find('.rc-field__clear').exists()).toBeFalsy();
    wrapper.setProps({ clearTrigger: 'always' });
    expect(wrapper.find('.rc-field__clear').exists()).toBeTruthy();
  });

  it('should render extra slot correctly', () => {
    wrapper = mount(<Field type="text" value="test" extra={<Icon name="success" />} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should change cell size when using size prop', () => {
    wrapper = mount(<Field type="text" value="test" size="large" />);
    expect(wrapper.find('.rc-cell--large').exists()).toBeTruthy();
  });

  it('should render label slot correctly', () => {
    wrapper = mount(<Field type="text" value="test" label="Lable" />);
    expect(toJson(wrapper.find('.rc-field__label'))).toMatchSnapshot();
  });

  it('should allow to set label width with unit', () => {
    wrapper = mount(<Field type="text" value="test" label="Lable" labelWidth="10rem" />);
    const label = wrapper.find('.rc-field__label');
    expect(label.getDOMNode().style.width).toEqual('10rem');
  });

  it('should allow to set label width without unit', () => {
    wrapper = mount(<Field type="text" value="test" label="Lable" labelWidth={100} />);
    const label = wrapper.find('.rc-field__label');
    expect(label.getDOMNode().style.width).toEqual('100px');
  });

  it('should render label class name when using labelClass prop', () => {
    wrapper = mount(<Field type="text" value="test" label="Lable" labelClass="custom-class" />);
    expect(wrapper.find('.rc-field__label').hasClass('custom-class')).toBeTruthy();
  });

  it('should change arrow direction when using arrow-direction prop', () => {
    wrapper = mount(<Field type="text" value="test" isLink arrowDirection="up" />);
    expect(wrapper.find('.van-icon-arrow-up').exists()).toBeTruthy();
  });

  it('should allow to format value with formatter prop', () => {
    const onChange = jest.fn();
    const formatter = (value) => value.replace(/\d/g, '');
    wrapper = mount(<Field type="text" formatter={formatter} onChange={onChange} />);
    wrapper
      .find('input')
      .simulate('change', { preventDefault: jest.fn(), target: { value: 'abc123' } });
    // expect(onChange).toHaveBeenCalledWith("abc") 功能同下;
    expect(onChange.mock.calls[0][0]).toEqual('abc');
  });

  it('should trigger format after blurring when format-trigger prop is blur', async () => {
    const onChange = jest.fn();
    const formatter = (value) => value.replace(/\d/g, '');
    wrapper = mount(
      <Field
        type="text"
        value="abc123"
        formatter={formatter}
        formatTrigger="onBlur"
        onChange={onChange}
      />,
    );

    wrapper
      .find('input')
      .simulate('change', { preventDefault: jest.fn(), target: { value: '123efg' } });
    expect(onChange.mock.calls[0][0]).toEqual('123efg');
    await wrapper.setProps({ value: '123efg' });

    wrapper.find('input').simulate('focus');
    wrapper.find('input').simulate('blur');
    expect(onChange).toHaveBeenCalledWith('efg');
  });

  it('should render word limit correctly', async () => {
    const onChange = jest.fn();
    wrapper = mount(<Field type="textarea" maxlength={3} showWordLimit onChange={onChange} />);
    wrapper.find('textarea').simulate('change', { target: { value: 'world' } });
    expect(onChange).toHaveBeenCalledWith('wor');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render word limit correctly when value is undefined', () => {
    wrapper = mount(<Field type="text" value={undefined} maxlength={3} showWordLimit />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render input name when using name prop', () => {
    wrapper = mount(<Field type="text" name="world" />);
    expect(wrapper.find('input').props().name).toEqual('world');
  });

  it('should render colon when using colon prop', () => {
    wrapper = mount(<Field type="text" label="world" colon />);
    expect(wrapper.find('.rc-field__label').text()).toEqual('world:');
    expect(toJson(wrapper.find('.rc-field__label'))).toMatchSnapshot();
  });

  it('should blur search input after pressing enter', async () => {
    const onKeypress = jest.fn();
    const inputRef = jest.fn();
    wrapper = mount(<Field getInputRef={inputRef} type="search" onKeypress={onKeypress} />);
    await wrapper.find('input').simulate('keypress');
    expect(onKeypress).toHaveBeenCalledTimes(1);
  });

  it('should format value after mounted if initial value is null', () => {
    wrapper = mount(<Field type="text" value="" />);
    expect(wrapper.find('input').props().value).toEqual('');
  });

  it('should change clear icon when using clear-icon prop', async () => {
    const onFocus = jest.fn();
    wrapper = mount(
      <Field type="text" clearIcon="cross" clearable value="test" onFocus={onFocus} />,
    );
    const input = wrapper.find('input');
    await input.simulate('focus');
    expect(onFocus).toHaveBeenCalled();
    expect(toJson(wrapper.find('.rc-field__clear'))).toMatchSnapshot();
  });

  it('should render error-message slot correctly', async () => {
    wrapper = mount(<Field type="text" errorMessage="this is an error message" />);
    expect(wrapper.find('.rc-field__error-message').text()).toEqual('this is an error message');
    expect(toJson(wrapper.find('.rc-field__error-message'))).toMatchSnapshot();
  });

  it('should render correctly when tooltip is not vaildElement or string ', async () => {
    wrapper = mount(<Field type="text" label="label" tooltip={{ message: 'world' }} />);
    const tooltip = wrapper.find('.rc-field__tooltip');
    await tooltip.simulate('click');
    const dialogMessage = global.document.body.querySelector('.rc-dialog__message');
    expect(dialogMessage.textContent).toEqual('world');
  });

  it('should emit onClear event correctly', async () => {
    const onChange = jest.fn();
    const onClear = jest.fn();

    wrapper = mount(
      <Field
        type="text"
        label="label"
        value="abc"
        clearable
        onClear={onClear}
        onChange={onChange}
      />,
    );
    await wrapper.find('input').simulate('focus');
    wrapper.find(Icon).simulate('touchstart');
    expect(onChange).toHaveBeenCalled();
    expect(onClear).toHaveBeenCalled();
  });

  it('should render intro correctly', async () => {
    wrapper = mount(<Field type="text" label="label" value="abc" intro={<Icon name="shop-o" />} />);
    expect(wrapper.find('i').hasClass('van-icon-shop-o')).toBeTruthy();
  });

  it('should render correctly when readonly', async () => {
    wrapper = mount(<Field type="text" label="label" value="abc" readonly />);
    wrapper.find('input').simulate('focus');
  });
});
