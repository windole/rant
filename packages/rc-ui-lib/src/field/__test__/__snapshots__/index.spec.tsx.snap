// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field should change clear icon when using clear-icon prop 1`] = `
Array [
  <Icon
    className="rc-field__clear"
    name="cross"
    onTouchStart={[Function]}
    size="16px"
    tag="i"
  >
    <Badge
      className="rc-field__clear van-icon van-icon-cross"
      onTouchStart={[Function]}
      showZero={true}
      style={
        Object {
          "color": undefined,
          "fontSize": "16px",
        }
      }
      tag="i"
    >
      <i
        className="rc-badge__wrapper rc-field__clear van-icon van-icon-cross"
        onTouchStart={[Function]}
        style={
          Object {
            "color": undefined,
            "fontSize": "16px",
          }
        }
      />
    </Badge>
  </Icon>,
  <Badge
    className="rc-field__clear van-icon van-icon-cross"
    onTouchStart={[Function]}
    showZero={true}
    style={
      Object {
        "color": undefined,
        "fontSize": "16px",
      }
    }
    tag="i"
  >
    <i
      className="rc-badge__wrapper rc-field__clear van-icon van-icon-cross"
      onTouchStart={[Function]}
      style={
        Object {
          "color": undefined,
          "fontSize": "16px",
        }
      }
    />
  </Badge>,
  <i
    className="rc-badge__wrapper rc-field__clear van-icon van-icon-cross"
    onTouchStart={[Function]}
    style={
      Object {
        "color": undefined,
        "fontSize": "16px",
      }
    }
  />,
]
`;

exports[`Field should render colon when using colon prop 1`] = `
<div
  className="rc-cell__title rc-field__label"
  style={Object {}}
>
  world
  :
</div>
`;

exports[`Field should render error-message slot correctly 1`] = `
<div
  className="rc-field__error-message"
>
  this is an error message
</div>
`;

exports[`Field should render extra slot correctly 1`] = `
<ForwardRef
  clearIcon="clear"
  clearTrigger="focus"
  extra={
    <Icon
      name="success"
      tag="i"
    />
  }
  formatTrigger="onChange"
  type="text"
  value="test"
>
  <Cell
    className="rc-field"
    extra={
      <Icon
        name="success"
        tag="i"
      />
    }
    icon={null}
    title={null}
    titleClass="rc-field__label"
    titleStyle={Object {}}
    valueClass="rc-field__value"
  >
    <div
      className="rc-cell rc-field"
    >
      <div
        className="rc-cell__value rc-cell__value--alone rc-field__value"
      >
        <div
          className="rc-field__body"
        >
          <input
            className="rc-field__control"
            onBlur={[Function]}
            onChange={[Function]}
            onCompositionEnd={[Function]}
            onCompositionStart={[Function]}
            onCompositionUpdate={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            placeholder=""
            type="text"
            value="test"
          />
        </div>
      </div>
      <Icon
        name="success"
        tag="i"
      >
        <Badge
          className="van-icon van-icon-success"
          showZero={true}
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
            }
          }
          tag="i"
        >
          <i
            className="rc-badge__wrapper van-icon van-icon-success"
            style={
              Object {
                "color": undefined,
                "fontSize": undefined,
              }
            }
          />
        </Badge>
      </Icon>
    </div>
  </Cell>
</ForwardRef>
`;

exports[`Field should render label slot correctly 1`] = `
<div
  className="rc-cell__title rc-field__label"
  style={Object {}}
>
  Lable
</div>
`;

exports[`Field should render textarea when type is textarea 1`] = `
<ForwardRef
  autosize={true}
  clearIcon="clear"
  clearTrigger="focus"
  formatTrigger="onChange"
  type="textarea"
>
  <Cell
    className="rc-field"
    icon={null}
    title={null}
    titleClass="rc-field__label"
    titleStyle={Object {}}
    valueClass="rc-field__value"
  >
    <div
      className="rc-cell rc-field"
    >
      <div
        className="rc-cell__value rc-cell__value--alone rc-field__value"
      >
        <div
          className="rc-field__body"
        >
          <textarea
            className="rc-field__control"
            onBlur={[Function]}
            onChange={[Function]}
            onCompositionEnd={[Function]}
            onCompositionStart={[Function]}
            onCompositionUpdate={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            placeholder=""
          />
        </div>
      </div>
    </div>
  </Cell>
</ForwardRef>
`;

exports[`Field should render word limit correctly 1`] = `
<ForwardRef
  clearIcon="clear"
  clearTrigger="focus"
  formatTrigger="onChange"
  maxlength={3}
  onChange={
    [MockFunction] {
      "calls": Array [
        Array [
          "wor",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  showWordLimit={true}
  type="textarea"
>
  <Cell
    className="rc-field rc-field--min-height"
    icon={null}
    title={null}
    titleClass="rc-field__label"
    titleStyle={Object {}}
    valueClass="rc-field__value"
  >
    <div
      className="rc-cell rc-field rc-field--min-height"
    >
      <div
        className="rc-cell__value rc-cell__value--alone rc-field__value"
      >
        <div
          className="rc-field__body"
        >
          <textarea
            className="rc-field__control rc-field__control--min-height"
            onBlur={[Function]}
            onChange={[Function]}
            onCompositionEnd={[Function]}
            onCompositionStart={[Function]}
            onCompositionUpdate={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            placeholder=""
          />
        </div>
        <div
          className="rc-field__word-limit"
        >
          <span
            className="rc-field__word-num"
          >
            0
          </span>
          /
          3
        </div>
      </div>
    </div>
  </Cell>
</ForwardRef>
`;

exports[`Field should render word limit correctly when value is undefined 1`] = `
<ForwardRef
  clearIcon="clear"
  clearTrigger="focus"
  formatTrigger="onChange"
  maxlength={3}
  showWordLimit={true}
  type="text"
>
  <Cell
    className="rc-field"
    icon={null}
    title={null}
    titleClass="rc-field__label"
    titleStyle={Object {}}
    valueClass="rc-field__value"
  >
    <div
      className="rc-cell rc-field"
    >
      <div
        className="rc-cell__value rc-cell__value--alone rc-field__value"
      >
        <div
          className="rc-field__body"
        >
          <input
            className="rc-field__control"
            onBlur={[Function]}
            onChange={[Function]}
            onCompositionEnd={[Function]}
            onCompositionStart={[Function]}
            onCompositionUpdate={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            placeholder=""
            type="text"
          />
        </div>
        <div
          className="rc-field__word-limit"
        >
          <span
            className="rc-field__word-num"
          >
            0
          </span>
          /
          3
        </div>
      </div>
    </div>
  </Cell>
</ForwardRef>
`;
